pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -489,21 +503,59 @@ func (bfs *Service) IsActive() bool {
 	return bfs.isActive.Load()
 }
 
+// retry function with exponential backoff.
+func (bfs *Service) retry(action func() error) error {
+	var (
+		err     error
+		backoff = initialBackoff
+	)
+
+	for range maxRetries {
+		if err = action(); err == nil {
+			return nil
+		}
+
+		if bfs.ctx.Err() != nil {
+			return bfs.ctx.Err()
+		}
+		timer := time.NewTimer(backoff)
+		select {
+		case <-timer.C:
+		case <-bfs.ctx.Done():
+			return bfs.ctx.Err()
+		}
+		timer.Stop()
```

Comment:
Stop/drain timer routine properly, read the documentation to `Reset` and `Stop`.